table/tables/partition.go

```diff
@@ -235,8 +242,8 @@ func generateHashPartitionExpr(tblInfo *model.TableInfo) (*PartitionExpr, error)
 }
 
 // PartitionExpr returns the partition expression.
-func (t *partitionedTable) PartitionExpr() *PartitionExpr {
-	return t.partitionExpr
+func (t *partitionedTable) PartitionExpr(columns []*expression.Column, names types.NameSlice) (*PartitionExpr, error) {
+	return newPartitionExprBySchema(t.meta, columns, names)
```

Comment:
Is there any way to avoid this allocation every time?

Reply:
Maybe we can traverse the Expression, find all columns and rewrite them. We can do it in future.